File: neat.js
class Neat {
    constructor(nInputs, nOutputs, clients) {
        this.clients = clients
        this.nInputs = nInputs
        this.nOutputs = nOutputs
        this.bias = 1
        this.neurons = []
        this.connections = []
        this.nNeurons = 0
        this.nConnections = 0

        for (let i = 0; i < nInputs; ++i) {
            this.neurons.push({innov: this.nNeurons, type: "input"})
            this.nNeurons++;
        }
        for (let i = 0; i < nOutputs; ++i) {
            this.neurons.push({innov: this.nNeurons, type: "output"})
            this.nNeurons++;
        }
        
        this.genoms = new Array(this.clients);
        for (let i = 0; i < this.clients; ++i) {
            this.genoms[i] = this.newGenom();
        }

    }

    connectionExists(n1, n2) {
        for (let i = 0; i < this.connections.length; ++i) {
            let c = this.connections[i];
            if (c.in == n1 && c.out == n2) {
                return {innov: c.innov, in: n1, out: n2, w: c.w, enabled: true}
            }
        }
        return false;
    } 

    neuronExists(innov) {
        for (let i = 0; i < this.neurons.length; ++i) {
            let c = this.neurons[i];
            if (c.innov == innov) {
                return {innov: c.innov, type: c.type}
            }
        }
        return false;
    }

    isCyclicRec(genom, i, visited, stack) {
        
    }

    isCyclic(genom, n1, n2) {
        let visited = new Array(genom.neurons.length);
        let stack = new Array(genom.neurons.length);
        visited.fill(true);
        stack.fill(true);
        genom.connections.push({
            innov: this.nConnections + 1, in: n1, out: n2, w: 0, enabled: true
        });

        for (let i = 0; i < genom.neurons.length; i++) {
            if (isCyclicRec(genom, i, visited, stack)) {
                genom.connections.pop();
                return true;
            }
        }

        genom.connections.pop();
        return false; 
    }
    
    // Comprova
    randomConnection(genom) {
        let n1 = Math.floor(Math.random()*genom.neurons.length);
        let n2 = Math.floor(Math.random()*genom.neurons.length);
        let malament = true;
        while(malament) {
            if (genom.neurons[n1].type == "output") {
                n1 = Math.floor(Math.random()*genom.neurons.length);
            }
            else if (genom.neurons[n2].type == "input") {
                n2 = Math.floor(Math.random()*genom.neurons.length);
            }
            else if (isCyclic(genom, n1, n2)) {
                n1 = Math.floor(Math.random()*genom.neurons.length);
                n2 = Math.floor(Math.random()*genom.neurons.length);
            }
            else malament = false;
        }
        let weight = Math.random();
        // Si ja existeix la conexió, la tornem
        let c = this.connectionExists(n1, n2);
        if (c) {
            return c
        }
        // Sino la tornem a crear
        let index = this.nConnections; 
        this.nConnections++;
        this.connections.push({
            innov: index, in: n1, out: n2, w: weight, enabled: true
        });
        return {innov: index, in: n1, out: n2, w: weight, enabled: true}
    }

    addConnection(n1, n2, w) {

        let c = this.connectionExists(n2, n2);
        if (c) {
            c.w = w;
            return c;
        }

        let index = this.nConnections; 
        this.nConnections++;

        this.connections.push({
            innov: index, in: n1, out: n2, w: w, enabled: true
        });
        return {innov: index, in: n1, out: n2, w: w, enabled: true}
    }

    addNeuron(innov) {
        let n = this.neuronExists(innov);
        if (n) {
            return n;
        }
        
        let index = this.nNeurons; 
        this.nNeurons++;
        this.neurons.push({innov: index, type: "hidden"});
        return {innov: index, type: "hidden"};
    }

    newGenom() {
        let newNeurons = [];
        for (let i = 0; i < this.neurons.length; ++i) {
            let newN = this.neurons[i];
            newNeurons.push({innov: newN.innov, type: newN.type})
        }
        return new Genom(this, newNeurons, this.nInputs, this.nOutputs);
    }
}

File: drift-v2/neat.py
from genom import Genom, Neuron, Connection
from species import Species
import random
import math

class Neat():
    def __init__(self, n_inp, n_out, clients, population):
        self.clients = clients
        self.population = population
        self.n_inputs = n_inp
        self.n_outputs = n_out
        self.bias = 1
        self.neurons = []
        self.connections = []
        self.n_neurons = 0
        self.n_connections = 0
        self.genoms = []
        self.species = []

        for i in range(n_inp):
            self.neurons.append(Neuron(self.n_neurons, "input"))
            self.n_neurons += 1
        
        for i in range(n_out):
            self.neurons.append(Neuron(self.n_neurons, "output"))
            self.n_neurons += 1
        
        for i in range(clients):
            new_neurons = [n.copy() for n in self.neurons]
            self.genoms.append(Genom(self, new_neurons, self.n_inputs, self.n_outputs))

    def connection_exists(self, n1, n2):
        for c in self.connections:
            if c.inp == n1 and c.out == n2:
                return c.copy()
        return False

    def neuron_exists(self, n_innov):
        for n in self.neurons:
            if n.innov == n_innov:
                return n.copy()
        return False

    def add_neuron(self, innov):
        n = self.neuron_exists(innov)
        if n: return n
        index = self.n_neurons
        self.n_neurons += 1
        self.neurons.append(Neuron(index, "hidden"))
        return Neuron(index, "hidden")
    
    def add_connection(self, n1, n2, w):
        c = self.connection_exists(n1, n2)
        if c:
            c.w = w
            return c

        index = self.n_connections
        self.n_connections += 1

        self.connections.append(Connection(index, n1, n2, w))
        return Connection(index, n1, n2, w)

    def reproduction(self, fitness_array):

        def check_species(g):
            for s in self.species:
                if s.check_if_member(g):
                    return True
            return False

        for i, g in enumerate(self.genoms):
            g.fitness = fitness_array[i]
            if not check_species(g):
                self.species.append(Species(g))

        n_extint = 0
        new_species = []
        for s in self.species:
            s.sort()
            extinct = s.kill()
            if extinct: 
                n_extint += 1
            else: 
                new_species.append(s.reproduce())

        self.species = new_species

        n_extra = self.clients - len(self.genoms)
        for i in range(n_extra):
            random.choice(self.genoms).reproduce_extra()

